src/Models/full_stokes/isothermal/kernels.jl

```diff
@@ -2,14 +2,24 @@ using KernelAbstractions
 
 using FastIce.GridOperators
 
-@kernel function update_η!(η, η_rh, χ, grid, fields, args...)
+@inline @generated function within(grid::CartesianGrid{N}, f::Field{T,N}, I::CartesianIndex{N}) where {T,N}
+    quote
+        Base.Cartesian.@nall $N i->I[i] <= size(grid, location(f, Val(i)), i)
+    end
+end
+
+@kernel function update_η!(η, η_rh, χ, grid, fields, offset=nothing)
     I = @index(Global, Cartesian)
-    ηt = η_rh(grid, I, fields, args...)
-    @inbounds η[I] = exp(log(η[I]) * (1 - χ) + log(ηt) * χ)
+    isnothing(offset) || (I += offset)
+    @inbounds begin
+        ηt = η_rh(grid, I, fields)
+        η[I] = exp(log(η[I]) * (1 - χ) + log(ηt) * χ)
+    end
 end
 
-@kernel function update_σ!(Pr, τ, V, η, Δτ, Δ)
+@kernel function update_σ!(Pr, τ, V, η, Δτ, Δ, offset=nothing)
     I = @index(Global, Cartesian)
+    isnothing(offset) || (I += offset)
     @inbounds if checkbounds(Bool, Pr, I)
```

Comment:
Should we use `within` here as well?

Reply:
Not really, `within` is rather fragile as of now, I guess I'd postpone wide application until I revamp `Fields` API a bit